### PR TITLE
Clean up relative path code

### PR DESCRIPTION
**Pull request by milleruntime**
* Drop methods in VolumeManager
* Create TabletFileUtil to do path validation

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+   * Validate if string is a valid path. Return same string or throw exception if not valid. This
+   * was added to facilitate more use of TabletFile over String but this puts the validation in one
+   * location in the case where TabletFile can't be used.
+   */
+  public static String validate(String path) {
+    Path p = new Path(path);
+    if (p.toUri().getScheme() == null) {
+      throw new IllegalArgumentException("Invalid path provided, no scheme in " + path);
+    }
+    return path;
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java]: Did we want to have this normalize the path also by returning the normalized version from `Path`?

File: server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java [hunk 2]
```diff
@@ -192,10 +192,9 @@ private BatchWriter createWriter(TableId tableId) {
     }
   }
 
-  public static Mutation createDeleteMutation(ServerContext context, TableId tableId,
-      String pathToRemove) {
-    Path path = context.getVolumeManager().getFullPath(tableId, pathToRemove);
-    Mutation delFlag = new Mutation(new Text(DeletesSection.encodeRow(path.toString())));
+  public static Mutation createDeleteMutation(String pathToRemove) {
+    String path = TabletFileUtil.validate(pathToRemove);
+    Mutation delFlag = new Mutation(new Text(DeletesSection.encodeRow(path)));
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java]: The old code did a kind of normalization here against the volume manager's filesystem. I would hope that by the time it gets here, all callers would be providing a normalized string, but care should be taken if the parameter is provided from user input, that we normalize it first. I didn't track down all the callers to check this.

File: server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java [hunk 3]
```diff
@@ -149,7 +150,8 @@ private static KeyExtent fixSplit(ServerContext context, TableId tableId, Text m
 
           for (Entry<Key,Value> entry : scanner3) {
             if (entry.getKey().compareColumnFamily(DataFileColumnFamily.NAME) == 0) {
-              origDatafileSizes.put(new FileRef(context.getVolumeManager(), entry.getKey()),
+              String tabletFilePath = entry.getKey().getColumnQualifierData().toString();
+              origDatafileSizes.put(new FileRef(TabletFileUtil.validate(tabletFilePath)),
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java]: This would be more readable if the validate call wrapped the right-hand-side of the assignment to `tabletFilePath` rather than being embedded into a constructor inside a parameter to the `put` method call.

File: server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java [hunk 4]
```diff
@@ -278,15 +278,12 @@ public static void addDeleteEntries(KeyExtent extent, Set<FileRef> datafilesToDe
     // TODO could use batch writer,would need to handle failure and retry like update does -
     // ACCUMULO-1294
     for (FileRef pathToRemove : datafilesToDelete) {
-      update(context,
-          ServerAmpleImpl.createDeleteMutation(context, tableId, pathToRemove.path().toString()),
-          extent);
+      update(context, ServerAmpleImpl.createDeleteMutation(pathToRemove.path().toString()), extent);
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/util/MetadataTableUtil.java]: An impl method could take the more precise type, rather than the `toString()` version, without affecting the API.

File: server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java (hunk 2 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java]: Also, there's a lot of calls to `toString()` on the path to validate. This is the only method that calls validate without calling `toString()` first. However, many of this method's own callers call `toString()` first on the parameter. I think some of this could be simplified if everybody stopped calling `toString()`, and we kept the more concrete type, and only call `toString()` inside the validate method, if necessary. It simplifies a lot of the calling code.

File: core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java]: I think that would be good.  You thinking return the ```Path``` object or ```p.toString()```?

File: server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java (hunk 3 above)
Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java]: I 100% wrote it this way for nice formatting as any other way wraps the line of code.  I think your way make sense logically though.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java [hunk 5]
```diff
@@ -419,66 +407,6 @@ public Path matchingFileSystem(Path source, String[] options) {
     return null;
   }
 
-  @Override
-  public Path getFullPath(TableId tableId, String path) {
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: There is another method in this class named `getVolumeByPath` that has relative path handling code.  Would that be something to address in this PR?  Or should that be another PR?

File: server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java [hunk 6]
```diff
@@ -319,9 +319,9 @@ public void delete(SortedMap<String,String> confirmedDeletes) throws TableNotFou
               // of deleting something that should not be deleted. Must not change value of delete
               // variable because thats whats stored in metadata table.
               log.debug("Volume replaced {} -> {}", delete, switchedDelete);
-              fullPath = fs.getFullPath(FileType.TABLE, switchedDelete);
+              fullPath = new Path(switchedDelete);
```
Comment (keith-turner) [inline server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java]: Should these be validated?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletData.java [hunk 7]
```diff
@@ -58,19 +58,18 @@ public TabletData(KeyExtent extent, VolumeManager fs, TabletMetadata meta) {
     this.flushID = meta.getFlushId().orElse(-1);
     this.directoryName = meta.getDirName();
     this.logEntries.addAll(meta.getLogs());
-    meta.getScans().forEach(tabletFile -> scanFiles
-        .add(new FileRef(fs, tabletFile.getMetadataEntry(), meta.getTableId())));
+    meta.getScans()
+        .forEach(tabletFile -> scanFiles.add(new FileRef(tabletFile.getMetadataEntry())));
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletData.java]: Would probably be useful to have a FileRef constructor that takes a TabletFile type.  This way its always converted easily and consistently.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java [hunk 8]
```diff
@@ -3217,7 +3216,7 @@ public void recover(VolumeManager fs, KeyExtent extent, List<LogEntry> logEntrie
     Collections.sort(sorted, (e1, e2) -> (int) (e1.timestamp - e2.timestamp));
     for (LogEntry entry : sorted) {
       Path recovery = null;
-      Path finished = RecoveryPath.getRecoveryPath(fs.getFullPath(FileType.WAL, entry.filename));
+      Path finished = RecoveryPath.getRecoveryPath(new Path(entry.filename));
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: Should this be validated?

File: server/master/src/main/java/org/apache/accumulo/master/TabletGroupWatcher.java [hunk 9]
```diff
@@ -614,7 +615,8 @@ private void deleteTablets(MergeInfo info) throws AccumuloException {
       for (Entry<Key,Value> entry : scanner) {
         Key key = entry.getKey();
         if (key.compareColumnFamily(DataFileColumnFamily.NAME) == 0) {
-          datafiles.add(new FileRef(this.master.fs, key));
+          datafiles
+              .add(new FileRef(TabletFileUtil.validate(key.getColumnQualifierData().toString())));
```
Comment (keith-turner) [inline server/master/src/main/java/org/apache/accumulo/master/TabletGroupWatcher.java]: With a method like the following, could possibly do `datafiles.add(new FileRef(TabletFile.from(key)));`.  Not completely sure about this, its just something I thought of while looking at this code.

```java
class TableFile {
 public static TableFile from(Key k){
   switch(k.getColumnFamilyData().toString()) {
     case SCAN:
     case DATA:
         return new TableFile(k.getColumnQualifierData().toString());
   }
 }
}
```

File: server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java (hunk 2 above)
Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java]: Which object do you think would be better?  The issue is that we have some use of Path, ByteSequence and String.  Calls to the ```createDeleteMutation()``` method are particularly hairy as there are a few spots where we parse the string and do string manipulation:
[gc](https://github.com/apache/accumulo/blob/3fd5cad92f9b63ac19e4466f3f2d5237b905262c/server/base/src/main/java/org/apache/accumulo/server/gc/GcVolumeUtil.java#L36)
[upgrader](https://github.com/apache/accumulo/blob/3276e788e8fa7ebf9c5f1fa4f37f49bad4eaeb3a/server/master/src/main/java/org/apache/accumulo/master/upgrade/Upgrader9to10.java#L434)

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java (hunk 8 above)
Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: I am not sure what we could validate with the WAL filenames.  But the filename is coming from LogEntry so I think we should do it when the LogEntry is created, not after the TabletServer gets the list of files.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java (hunk 5 above)
Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManagerImpl.java]: I think ```getVolumeByPath``` may have to be done in another change, related to the DEFAULT volume code.  That method returns the DEFAULT volume if it detects a relative path, which I could remove but I wanted to look more into whether the DEFAULT volume can be removed with the other legacy properties.

File: server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java (hunk 6 above)
Comment (milleruntime) [reply server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java]: Looking at the code, I can't figure out if these are delete markers or the files to delete...

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java (hunk 8 above)
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: > so I think we should do it when the LogEntry is created

I like that approach.  We do need to avoid constantly validating the same data over and over in the code.  I think the goal should be that data is only validated after its read from a persistent store and no where else.

File: server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java (hunk 2 above)
Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java]: I think you are pointing to an overall problem with a lot of the internal code.  One of my goals with the new ```TabletFile``` (similar to how we are trying to use Ample as much as possible) was to try and replace as much overuse of String and FileRef as we can.  We have a ```LogEntry``` which I think is good for WALs but we might need some more internal types for strings that aren't TabletFiles.

Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java]: That being said, I think a lot of that can be done in other PRs while the goal of this one was to remove some of the relative path code that was no longer needed in VolumeManager

File: core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java (hunk 1 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java]: Not sure. Probably Path, but it depends on caller needs. I haven't investigated.

File: server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java (hunk 2 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/metadata/ServerAmpleImpl.java]: I'm happy to defer to future improvements, if what can be done here is limited.

File: server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java (hunk 3 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java]: LOL, I wasn't thinking about formatting. Now, I'm not so sure. It might fit if `validate` were overloaded to take a `Text` so you didn't have to do `toString()`? Whatever you think best.

Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java]: Ah man... I don't like overloaded methods to begin with, let alone adding one purely for formatting haha.  I'll see how your suggestion fits and if its ugly ill just keep what I had.

Comment (milleruntime) [reply server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java]: Either way, I think this could get replaced with ```TabletFile``` or ```Ample``` eventually.

File: server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java (hunk 6 above)
Comment (keith-turner) [reply server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java]: These are delete markers that have no references and are therefore safe to delete.  The code after this deletes the files and then deletes the markers from the metadata table.

File: server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java (hunk 3 above)
Comment (ctubbsii) [reply server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java]: It's not purely formatting. I would argue that the overloaded method would help centralize the logic related to validation by simplifying calling code, and that, in turn, makes it easier to swap out later (because working with concrete types rather than Strings helps you create replacement code that satisfies all your caller use cases). Also, avoiding the `toString()` calls and using the concrete types instead, can help find a lot of situations where code is being unnecessarily converted back and forth to/from String.

File: core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/metadata/TabletFileUtil.java]: I think these could also get replaced with TabletFile or Ample as well.  So I will just leave as String.  If we can replace them all, this class will go away

**Comment by milleruntime** [comment]
I am going to merge this since most of the comments pertained to future work.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java (hunk 8 above)
Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java]: I checked and ```LogEntry.fromKeyValue(key, val)``` does some checking. And this is what is called when the data is read from the metadata in ```TabletMetadata```
